execution/executor-types/src/in_memory_state_calculator.rs

```diff
@@ -276,51 +263,41 @@ impl InMemoryStateCalculator {
 // Updates the `state_cache` to reflect the latest value.
 // Returns all state keys touched.
 pub fn process_write_set(
-    transaction: Option<&Transaction>,
+    transaction: &Option<&Transaction>,
     state_cache: &mut HashMap<StateKey, StateValue>,
     write_set: WriteSet,
-) -> Result<HashSet<StateKey>> {
+) -> Result<HashMap<StateKey, StateValue>> {
     // Find all keys this transaction touches while processing each write op.
-    let mut updated_keys = HashSet::new();
-    for (state_key, write_op) in write_set.into_iter() {
-        process_state_key_write_op(
-            transaction,
-            state_cache,
-            &mut updated_keys,
-            state_key,
-            write_op,
-        )?;
-    }
-
-    Ok(updated_keys)
+    write_set
+        .into_iter()
+        .map(|(state_key, write_op)| {
+            process_state_key_write_op(transaction, state_cache, state_key, write_op)
+        })
+        .collect::<Result<_>>()
 }
 
 fn process_state_key_write_op(
-    transaction: Option<&Transaction>,
+    transaction: &Option<&Transaction>,
     state_cache: &mut HashMap<StateKey, StateValue>,
-    updated_keys: &mut HashSet<StateKey>,
     state_key: StateKey,
     write_op: WriteOp,
-) -> Result<()> {
+) -> Result<(StateKey, StateValue)> {
+    let state_value = match write_op {
+        WriteOp::Value(new_value) => StateValue::from(new_value),
+        WriteOp::Deletion => StateValue::empty(),
+    };
     match state_cache.entry(state_key.clone()) {
         hash_map::Entry::Occupied(mut entry) => {
-            match write_op {
-                WriteOp::Value(new_value) => entry.insert(StateValue::from(new_value)),
-                WriteOp::Deletion => entry.insert(StateValue::empty()),
-            };
+            entry.insert(state_value.clone());
```

Comment:
irrelevant: we might want to make state values light to clone (use an internal `Arc` for example; similar to the argument about we are cloning jmt nodes too much. @grao1991 )